sources/rest_api/client.py

```diff
@@ -0,0 +1,96 @@
+from typing import Optional, Dict, Any, Generator
+import copy
+from requests.auth import AuthBase
+
+from dlt.sources.helpers import requests
+
+from .paginators import BasePaginator, UnspecifiedPaginator
+from .detector import create_paginator
+
+from .utils import join_url
+
+class RESTClient:
+    """A generic REST client for making requests to an API.
+
+    Attributes:
+        base_url (str): The base URL of the API.
+        headers (Optional[Dict[str, str]]): Headers to include in all requests.
+        auth (Optional[AuthBase]): An authentication object to use for all requests.
+        paginator (Optional[BasePaginator]): A paginator object for handling API pagination.
+            Note that this object will be deepcopied for each request to ensure that the
+            paginator state is not shared between requests.
+    """
+
+    def __init__(
+        self,
+        base_url: str,
+        headers: Optional[Dict[str, str]] = None,
+        auth: Optional[AuthBase] = None,
```

Comment:
I think typing for this should be as `Optional[Type[AuthBase]]` or `AuthBase[T]` and probably same for pagination abstractions